Data/HashMap/Internal/Debug.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE BangPatterns     #-}
+{-# LANGUAGE CPP              #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- | = WARNING
+--
+-- This module is considered __internal__.
+--
+-- The Package Versioning Policy __does not apply__.
+--
+-- The contents of this module may change __in any way whatsoever__
+-- and __without any warning__ between minor versions of this package.
+--
+-- Authors importing this module are expected to track development
+-- closely.
+--
+-- = Description
+--
+-- Debugging utilities for 'HashMap's.
+
+module Data.HashMap.Internal.Debug
+    ( valid
+    , Validity(..)
+    , Error(..)
+    , SubHash
+    , SubHashPath
+    ) where
+
+import Data.Bits             (complement, countTrailingZeros, popCount,
+                              unsafeShiftL, unsafeShiftR, (.&.))
+import Data.Hashable         (Hashable)
+import Data.HashMap.Internal (Bitmap, Hash, HashMap (..), Leaf (..),
+                              bitsPerSubkey, fullBitmap, hash,
+                              isLeafOrCollision, maxChildren, sparseIndex,
+                              subkeyMask)
+import Data.Semigroup        (Sum (..))
+
+import qualified Data.HashMap.Internal.Array as A
+
+
+#if !MIN_VERSION_base(4,11,0)
+import Data.Semigroup (Semigroup (..))
+#endif
+
+data Validity k = Invalid (Error k) SubHashPath | Valid
+  deriving (Eq, Show)
+
+instance Semigroup (Validity k) where
+  Valid <> y = y
+  x     <> _ = x
+
+instance Monoid (Validity k) where
+  mempty = Valid
+  mappend = (<>)
+
+-- | An error corresponding to a broken invariant.
+--
+-- See 'HashMap' for the documentation of the invariants.
+data Error k
+  = INV1_internal_Empty
+  | INV2_bad_BitmapIndexed_size !Int
+  | INV3_bitmap_array_size_mismatch !Bitmap !Int
+  | INV4_BitmapIndexed_invalid_single_subtree
+  | INV5_misplaced_hash !Hash
+  | INV6_key_hash_mismatch k !Hash
+  | INV7_bad_Full_size !Int
+  | INV8_Collision_size !Int
+  | INV9_Collision_duplicate_key k !Hash
```

Comment:
What are the invariant numbers for?

Reply:
Actually I'm not sure whether these will be useful. They are just reference numbers that can hopefully stay constant when we reword the invariants or rename the error constructors.